### DISCUSSION TITLE
Matcher is only as good as its last matching

### DISCUSSION THREAD

**Discussion by obakamai** [etorreborre/specs2#1206] [Q&A] [answered]
```scala
object AXTests extends Specification {

  def is: SpecStructure = s2"""
          test assertions      $e1
  """
 
  def e1: MatchResult[Any] = {
    val t = 1
    val r = 3
    t must beEqualTo(4) //ignores the MatchFailure
    r must beEqualTo(3)
  }
}
```
 
This will pass as it seems the matcher will ignore the first match failure (the equals to 4) and will pass the test cause the last one match. What Am I missing here ?

**Comment by etorreborre** [comment]
Hi @obakamai. You're correct! By default, specs2 is a _functional_ library. At its core no exceptions are thrown.
When you write `t must beEqualTo(4)` you get back a `Result` which is just a normal value. 
This was done on purpose in order to:

 - incite people to get into a functional mindset 
 - allow an easier reuse of the core library because composing code with exceptions can be difficult

This means that the default way to check 2 expectations in a default specification is to write:
```scala
(t must beEqualTo(4)) and 
(r must beEqualTo(3))
```
You can also reuse the fact that `Result` are values with a `Monoid` instance, so you can "sum" them by putting them in a list:
```scala
import org.specs2.fp.syntax.*

List(
  1 must beEqualTo(1),
  2 must beEqualTo(4),
  3 must beEqualTo(3),
).sumAll
```

This will return:
```sh
[info]  x e2
[error]  2 != 4
```

You might say that it is wasteful because this will evaluate all the assertions. But this can be an advantage too! Because, with a different `Monoid` instance for `Result` you can collect all the failures at once with the right imports:
```scala
import org.specs2.fp.*
import org.specs2.fp.syntax.*
import org.specs2.execute.*

given Monoid[Result] = Result.ResultFailuresMonoid(separator = "\n")

List(
  1 must beEqualTo(2),
  2 must beEqualTo(2),
  3 must beEqualTo(4),
).sumAll
```

This will print:
```sh
[error]  1 != 2
[error]  3 != 4
```

You can also go back to a more traditional style of writing assertions, with a short-circuiting behaviour. 

In order to do this, extend your specification with the `org.specs2.matcher.ThrownExpectations` trait. Now an exception will be thrown when you write `1 must beEqualTo(4)`.

**Comment by obakamai** [comment]
Hi @etorreborre thanks for you reply, 
Can you write your examples in Scala please (2.12; specs2 4.16.1) ? it seems there is a different implementation
Thanks!

**Reply by tabdulradi** [reply, marked as answer]
That's Scala 3 syntax. To make it Scala 2 replace `*` with `_` and `given Monoid[Result]` with `implicit val resultMonoidInstance: Monoid[Result]`
